### ISSUE TITLE
15.0.2: implementation vs. compile - implementation configuration is ignored

### ISSUE THREAD

**Issue by MarcelBruch** [nebula-plugins/gradle-lint-plugin#260]
When running 15.0.2 on my code base, gradle-lint seems to ignore dependencies made with 'implementation...'. Is there a way to use gradle-lint with more modern configurations?

Example from lint.patch:

diff --git a/java/intellij/modules/recommenders-java-statics/build.gradle b/java/intellij/modules/recommenders-java-statics/build.gradle
--- a/java/intellij/modules/recommenders-java-statics/build.gradle
+++ b/java/intellij/modules/recommenders-java-statics/build.gradle
@@ -34,6 +34,9 @@
     implementation 'org.eclipse.recommenders:org.eclipse.recommenders.utils:2.5.5-SNAPSHOT'
+    compile 'org.eclipse.recommenders:org.eclipse.recommenders.utils:2.5.5-SNAPSHOT'
 }

The patch simply adds a new line with 'compile..' to the existing file - for every dependency. I thus end up with a plenty of 'duplicates'.

Using deprecated-dependency-configuration to "fix" this in a second run also does not work because it simple replaces compile with implementation - but does not check whether there is already the same dependency implementation scope :-(

Any idea how to get this work?

**Comment by rpalcolea** [comment]
Hi, we modified the rule, so it should show you warnings properly in 15.0.3

We decided to not write changes to the build file with this rule as we don't know the real intent from the project, for example if you should be using `api` or `implementation` when applying `java-library`. That would require library author to decide accordingly.

With that in mind, the rule will show warnings with usage of old configurations and what should be changed with a potential suggestion.

Please give a try to `15.0.3`. You shouldn't have duplicated dependencies now

**Comment by MarcelBruch** [comment]
Thx! I tried 15.0.3. and noticed the warnings.

However, the root problem is still there: gw fixGradleLint -PgradleLint.rules=undeclared-dependency produces redundant dependencies:

Here's an example:
**Before:**

dependencies {
    implementation 'org.slf4j:slf4j-api:1.7.28'
    implementation 'org.springframework.boot:spring-boot-autoconfigure:2.1.8.RELEASE'
    implementation 'org.springframework.boot:spring-boot:2.1.8.RELEASE'
    implementation 'org.springframework.shell:spring-shell-standard-commands:2.0.1.RELEASE'
    implementation 'org.springframework.shell:spring-shell-standard:2.0.1.RELEASE'
    implementation 'org.springframework.shell:spring-shell-starter'
    implementation 'org.springframework:spring-context:5.1.9.RELEASE'
...
}

**after:**

dependencies {
    implementation 'org.slf4j:slf4j-api:1.7.28'
    implementation 'org.springframework.boot:spring-boot-autoconfigure:2.1.8.RELEASE'
    implementation 'org.springframework.boot:spring-boot:2.1.8.RELEASE'
    implementation 'org.springframework.shell:spring-shell-standard-commands:2.0.1.RELEASE'
    implementation 'org.springframework.shell:spring-shell-standard:2.0.1.RELEASE'
    implementation 'org.springframework:spring-context:5.1.9.RELEASE'
...
    compile 'org.slf4j:slf4j-api:1.7.28'
    compile 'org.springframework.boot:spring-boot-autoconfigure:2.1.8.RELEASE'
    compile 'org.springframework.boot:spring-boot:2.1.8.RELEASE'
    compile 'org.springframework.shell:spring-shell-standard-commands:2.0.1.RELEASE'
    compile 'org.springframework.shell:spring-shell-standard:2.0.1.RELEASE'
    compile 'org.springframework:spring-context:5.1.9.RELEASE'
}

My expectation is, that _gradle-lint undeclared-dependency_ understands that the dependencies declared in _implementation_  are not missing - and thus would not add the _compile_ lines . Does that make sense to you?

**Comment by rpalcolea** [comment]
Hi, this is interesting. I tried with with the latest in https://github.com/rpalcolea/unused-dependencies-example

```
❯ ./gradlew build

> Task :autoLintGradle

This project contains lint violations. A complete listing of the violations follows.
Because none were serious, the build's overall status was unaffected.

warning   unused-dependency                  this dependency is unused and can be removed
build.gradle:11
implementation 'org.slf4j:slf4j-api:1.7.28'

warning   unused-dependency                  this dependency is unused and can be removed
build.gradle:12
implementation 'org.springframework.boot:spring-boot-autoconfigure:2.1.8.RELEASE'

warning   unused-dependency                  this dependency is unused and can be removed
build.gradle:13
implementation 'org.springframework.boot:spring-boot:2.1.8.RELEASE'

✖ 3 problems (0 errors, 3 warnings)

To apply fixes automatically, run fixGradleLint, review, and commit the changes.
``` 

When I run `./gradlew fixLintGradle`

```
> Task :fixLintGradle

This project contains lint violations. A complete listing of my attempt to fix them follows. Please review and commit the changes.

fixed          unused-dependency                  this dependency is unused and can be removed
build.gradle:11
implementation 'org.slf4j:slf4j-api:1.7.28'

fixed          unused-dependency                  this dependency is unused and can be removed
build.gradle:12
implementation 'org.springframework.boot:spring-boot-autoconfigure:2.1.8.RELEASE'

fixed          unused-dependency                  this dependency is unused and can be removed
build.gradle:13
implementation 'org.springframework.boot:spring-boot:2.1.8.RELEASE'

Corrected 3 lint problems

``` 

and then the changes are to remove the unused dependencies, not duplicated:

```
plugins {
  id "nebula.lint" version "15.0.3"
  id "java"
}

repositories {
  jcenter()
}

dependencies  {
}

gradleLint {
  rules  = ['unused-dependency']
}

```

Any chance you could provide a reproducible example?
